webapp/src/Serializer/Shadowing/EventDenormalizer.php

```diff
@@ -0,0 +1,115 @@
+use Symfony\Component\Serializer\Normalizer\DenormalizerInterface;
+use Symfony\Component\Serializer\SerializerAwareInterface;
+use Symfony\Component\Serializer\SerializerAwareTrait;
+
+/**
+ * This class is responsible for changing an event array, either in the legacy 2020-03 format
+ * or the new 2022-07 format, into an Event object.
+ */
+class EventDenormalizer implements DenormalizerInterface, SerializerAwareInterface
+{
+    use SerializerAwareTrait;
+
+    /**
+     * @param array{id?: string,
+     *     type: string,
+     *     op?: string,
+     *     token?: string,
+     *     data: array{id: string}|array<array{id: string}>|null
+     * }                                  $data
+     * @param array{api_version?: string} $context
+     * @return Event<EventData>
+     *
+     * @throws ExceptionInterface
+     */
+    public function denormalize(
+        mixed $data,
+        string $type,
+        ?string $format = null,
+        array $context = []
+    ): Event {
+        if (!$this->supportsDenormalization($data, $type, $format, $context)) {
+            throw new InvalidArgumentException('Unsupported data.');
+        }
+
+        if (!$this->serializer instanceof DenormalizerInterface) {
+            throw new LogicException('Cannot denormalize attribute "data" because the injected serializer is not a denormalizer.');
+        }
+
+        $eventType = EventType::fromString($data['type']);
+        if ($this->getEventFeedFormat($data, $context) === EventFeedFormat::Format_2022_07) {
+            $operation = !isset($data['data']) ? Operation::DELETE : Operation::CREATE;
+            if (!isset($data['data'][0])) {
+                $data['data'] = [$data['data']];
+            }
+            $id = $operation === Operation::DELETE ? $data['id'] : $data['data'][0]['id'] ?? null;
+            return new Event(
+                $data['token'] ?? null,
+                $eventType,
+                $operation,
+                $id,
+                $this->serializer->denormalize($data['data'], EventData::class . '[]', $format, $context + ['event_type' => $eventType]),
+            );
+        } else {
+            return new Event(
+                $data['id'] ?? null,
+                $eventType,
+                Operation::from($data['op']),
+                $data['data']['id'],
+                [$this->serializer->denormalize($data['data'], EventData::class, $format, $context + ['event_type' => $eventType])],
+            );
+        }
+    }
+
+    /**
+     * @param array{api_version?: string} $context
+     */
+    public function supportsDenormalization(
+        mixed $data,
+        string $type,
+        ?string $format = null,
+        array $context = []
+    ): bool {
+        if (!is_array($data)) {
+            return false;
+        }
+        if ($type !== Event::class) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * @return array<string, bool|null>
```

Comment:
```suggestion
     * @return array{0: bool, 1: null}
```
Looks shorter.

Reply:
But also wrong, since the keys are strings, not integers.